capa/features/extractors/elffile.py

```diff
@@ -0,0 +1,153 @@
+    symbol_tables = [(idx, s) for idx, s in enumerate(elf.iter_sections()) if isinstance(s, SymbolTableSection)]
+
+    for section_index, section in symbol_tables:
+        if not isinstance(section, SymbolTableSection):
+            continue
+
+        if section["sh_entsize"] == 0:
+            logger.debug("Symbol table '%s' has a sh_entsize of zero!" % (section.name))
+            continue
+
+        logger.debug("Symbol table '%s' contains %s entries:" % (section.name, section.num_symbols()))
+
+        for nsym, symbol in enumerate(section.iter_symbols()):
+            if symbol.name and symbol.entry.st_info.type == "STT_FUNC":
+                # TODO symbol address
+                # TODO symbol version info?
+                yield Import(symbol.name), 0x0
+
+
+def extract_file_section_names(elf, **kwargs):
+    for section in elf.iter_sections():
+        if section.name:
+            yield Section(section.name), section.header.sh_addr
+        elif section.is_null():
+            yield Section("NULL"), section.header.sh_addr
+
+
+def extract_file_strings(buf, **kwargs):
+    yield from capa.features.extractors.common.extract_file_strings(buf)
+
+
+def extract_file_os(elf, buf, **kwargs):
+    # our current approach does not always get an OS value, e.g. for packed samples
+    # for file limitation purposes, we're more lax here
+    try:
+        os = next(capa.features.extractors.common.extract_os(buf))
+        yield os
+    except StopIteration:
+        yield OS("unknown"), 0x0
+
+
+def extract_file_format(**kwargs):
+    yield Format(FORMAT_ELF), 0x0
+
+
+def extract_file_arch(elf, **kwargs):
+    # TODO merge with capa.features.extractors.elf.detect_elf_arch()
+    arch = elf.get_machine_arch()
+    if arch == "x86":
+        yield Arch(ElfArch.I386), 0x0
+    elif arch == "x64":
+        yield Arch(ElfArch.AMD64), 0x0
+    else:
+        logger.warning("unsupported architecture: %s", arch)
+
+
+def extract_file_features(elf, buf):
+    """
+    extract file features from given sample
+
+    args:
+      elf (elftools.elf.elffile.ELFFile): the parsed ELFFile
+      buf: the raw sample bytes
+
+    yields:
+      Tuple[Feature, VA]: a feature and its location.
+    """
+
+    for file_handler in FILE_HANDLERS:
+        for feature, va in file_handler(elf=elf, buf=buf):
+            yield feature, va
+
+
+FILE_HANDLERS = (
+    # TODO extract file export names
+    # extract_file_export_names,
+    extract_file_import_names,
+    extract_file_section_names,
+    extract_file_strings,
+    # elffile doesn't have library matching
+    # extract_file_function_names,
+    extract_file_os,
+    extract_file_format,
+    extract_file_arch,
+)
+
+
+class ElfFeatureExtractor(FeatureExtractor):
+    def __init__(self, path: str):
+        super(ElfFeatureExtractor, self).__init__()
+        self.path = path
+        # TODO close where/when?
```

Comment:
```python
        with open(self.path, "rb") as f:
            self.elf = ELFFile(f)
```

results in 
```
Traceback (most recent call last):
  File "..venv3\Scripts\capa-script.py", line 33, in <module>
    sys.exit(load_entry_point('flare-capa', 'console_scripts', 'capa')())
  File "..capa\capa\main.py", line 968, in main
    pure_file_capabilities, _ = find_file_capabilities(rules, file_extractor, {})
  File "..capa\capa\main.py", line 116, in find_file_capabilities
    for feature, va in extractor.extract_file_features():
  File "..capa\capa\features\extractors\elffile.py", line 130, in extract_file_features
    for feature, va in extract_file_features(self.elf, buf):
  File "..capa\capa\features\extractors\elffile.py", line 93, in extract_file_features
    for feature, va in file_handler(elf=elf, buf=buf):
  File "..capa\capa\features\extractors\elffile.py", line 24, in extract_file_import_names
    symbol_tables = [(idx, s) for idx, s in enumerate(elf.iter_sections()) if isinstance(s, SymbolTableSection)]
  File "..capa\capa\features\extractors\elffile.py", line 24, in <listcomp>
    symbol_tables = [(idx, s) for idx, s in enumerate(elf.iter_sections()) if isinstance(s, SymbolTableSection)]
  File "..venv3\lib\site-packages\elftools\elf\elffile.py", line 133, in iter_sections
    yield self.get_section(i)
  File "..venv3\lib\site-packages\elftools\elf\elffile.py", line 112, in get_section
    section_header = self._get_section_header(n)
  File "..venv3\lib\site-packages\elftools\elf\elffile.py", line 526, in _get_section_header
    return struct_parse(
  File "..venv3\lib\site-packages\elftools\common\utils.py", line 39, in struct_parse
    stream.seek(stream_pos)
ValueError: seek of closed file
```

Reply:
1. could make this (single) extractor class a contextmanager and wrap the inner `f`
2. could use `io.BytesIO` of the full buffer in memory (assuming not too large)

Reply:
11644cbc312aa3866abe9302c0bde4e00776ab9e